dbms/src/Storages/StorageKafka.cpp

```diff
@@ -312,21 +305,85 @@ void StorageKafka::shutdown()
 
 void StorageKafka::updateDependencies()
 {
-    cancel_event.set();
+    event_update.set();
+}
+
+
+void StorageKafka::consumerConfiguration(struct rd_kafka_conf_s * conf)
+{
+    std::vector<char> errstr(512);
+
+    LOG_TRACE(log, "Setting brokers: " << brokers);
+    if (rd_kafka_conf_set(conf, "metadata.broker.list", brokers.c_str(), errstr.data(), errstr.size()) != RD_KAFKA_CONF_OK)
+        throw Exception(String(errstr.data()), ErrorCodes::INCORRECT_DATA);
+
+    LOG_TRACE(log, "Setting Group ID: " << group << " Client ID: clickhouse");
+
+    if (rd_kafka_conf_set(conf, "group.id", group.c_str(), errstr.data(), errstr.size()) != RD_KAFKA_CONF_OK)
+        throw Exception(String(errstr.data()), ErrorCodes::INCORRECT_DATA);
+
+    if (rd_kafka_conf_set(conf, "client.id", VERSION_FULL, errstr.data(), errstr.size()) != RD_KAFKA_CONF_OK)
+        throw Exception(String(errstr.data()), ErrorCodes::INCORRECT_DATA);
+
+    // We manually commit offsets after a stream successfully finished
+    rd_kafka_conf_set(conf, "enable.auto.commit", "false", nullptr, 0);
+
+    // Update consumer configuration from the configuration
+    const auto & config = context.getConfigRef();
+    if (config.has(CONFIG_PREFIX))
+        loadFromConfig(conf, config, CONFIG_PREFIX);
+
+    // Update consumer topic-specific configuration
+    for (const auto & topic : topics)
+    {
+        const auto topic_config_key = CONFIG_PREFIX + "_" + topic;
+        if (config.has(topic_config_key))
+            loadFromConfig(conf, config, topic_config_key);
+    }
+}
+
+
+StorageKafka::ConsumerPtr StorageKafka::claimConsumer(long wait_ms)
+{
+    // Wait for the first free consumer
+    if (wait_ms >= 0)
+    {
+        if (!semaphore.tryWait(wait_ms))
+            return nullptr;
+    }
+    else
+        semaphore.wait();
+
+    // Take the first available consumer from the list
+    std::lock_guard<std::mutex> lock(mutex);
+    auto consumer = consumers.back();
```

Comment:
Is it guaranteed that there are at least one available consumer at this point?

Reply:
This should be guaranteed by the semaphore. There are as many semaphore slots as there are consumers. If the code reaches this line, it means that semaphore reserved at least one slot for this caller (otherwise it would throw an exception or returned nullptr).